File: algorithms/lib/heap_sort.rb
require_relative "heap"

class Array
  def heap_sort!
    start = Time.new.nsec
    heap = BinaryMinHeap.new

    self.each { |el| heap.push(el)}

    i = 0
    until heap.store.empty?
      self[i] = heap.extract
      i += 1
    end
    p Time.new.nsec - start
    self
  end
end

File: algorithms/heapJS/heap_sort.js
let BinaryMinHeap = require("./heap");

function heapSort(array) {
  let heap = new BinaryMinHeap;

  while (array.length) {
    heap.push(array.pop())
  }

  let result = [];

  while (heap.count() > 0) {
    let el = heap.extract()
    
    result.push(el);
  }

  return result;
}

module.exports = heapSort;
